File: api/utlis/geo.py
import math
from typing import Callable, Dict, List, Tuple, TypeVar

T = TypeVar("T")  # Generic type for model


def haversine_distance_between_points(
    lat1: float, lon1: float, lat2: float, lon2: float
) -> float:
    R = 6378  # Radius of the Earth in kilometers
    p1 = math.radians(lat1)
    p2 = math.radians(lat2)
    delta_lambda = math.radians(lon2 - lon1)
    d = (
        math.acos(
            math.sin(p1) * math.sin(p2)
            + math.cos(p1) * math.cos(p2) * math.cos(delta_lambda)
        )
        * R
    )
    return d


def cosine_distance_between_points(
    lat1: float, lon1: float, lat2: float, lon2: float
) -> float:
    R = 6371000  # Radius of the Earth in meters
    p1 = math.radians(lat1)
    p2 = math.radians(lat2)
    delta_phi = p2 - p1
    delta_lambda = math.radians(lon2 - lon1)
    a = (
        math.sin(delta_phi / 2) ** 2
        + math.cos(p1) * math.cos(p2) * math.sin(delta_lambda / 2) ** 2
    )
    d = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a)) * R
    return d


def find_nearest(
    locations: List[T],
    reference: Dict[str, float],
    df: Callable[
        [float, float, float, float], float
    ] = haversine_distance_between_points,
    location_extract_method: Callable[[T], Tuple[float, float]] = lambda loc: (
        loc["lat"],
        loc["lon"],
    ),
) -> Tuple[T, float]:
    min_dist = float("inf")
    nearest = None

    for loc in locations:
        lat, lon = location_extract_method(loc)
        dist = df(reference["lat"], reference["lon"], lat, lon)
        if dist < min_dist:
            min_dist = dist
            nearest = loc

    return nearest, min_dist

File: web/src/utils/geo.js
 
export function haversineDistanceBetweenPoints(lat1, lon1, lat2, lon2) {
  const R = 6378; // km
  const p1 = lat1 * Math.PI/180;
  const p2 = lat2 * Math.PI/180;
  const deltaLon = lon2 - lon1;
  const deltaLambda = (deltaLon * Math.PI) / 180;
  const d = Math.acos(
    Math.sin(p1) * Math.sin(p2) + Math.cos(p1) * Math.cos(p2) * Math.cos(deltaLambda),
  ) * R;
  return d;
}

export function cosineDistanceBetweenPoints(lat1, lon1, lat2, lon2) {
  const R = 6371e3;
  const p1 = lat1 * Math.PI/180;
  const p2 = lat2 * Math.PI/180;
  const deltaP = p2 - p1;
  const deltaLon = lon2 - lon1;
  const deltaLambda = (deltaLon * Math.PI) / 180;
  const a = Math.sin(deltaP/2) * Math.sin(deltaP/2) +
            Math.cos(p1) * Math.cos(p2) *
            Math.sin(deltaLambda/2) * Math.sin(deltaLambda/2);
  const d = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)) * R;
  return d;
}


export function findNearest(locations, reference, df) {
  let minDist = Infinity;
  let nearest = null;
  for (const loc of locations) {
    const dist = df(reference.lat, reference.lon, loc.lat, loc.lon);

    if (dist < minDist) {
      minDist = dist;
      nearest = loc;
    }
  }
  
  return {nearest,minDist};
}